File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        def helper(nums: List[int]) -> int:
            
            n = len(nums)
            
            rob1 = 0
            rob2 = 0
            
            for r in nums:
                temp = max(rob1 + r, rob2)
                rob1 = rob2
                rob2 = temp
                
            return rob2
        
        if len(nums) == 1: return nums[0]
        
        return max(helper(nums[:-1:]), helper(nums[1::]))
            

File: 0213-house-robber-ii/0213-house-robber-ii.ts
function rob(nums: number[]): number {
    const n = nums.length
    if (n === 1) return nums[0]
    return Math.max(helper(nums.slice(0, n-1)), helper(nums.slice(1)))
    
    function helper(nums: number[]): number {
        let rob1 = 0
        let rob2 = 0
        for (const r of nums) {
            const temp = Math.max(rob1 + r, rob2)
            rob1 = rob2
            rob2 = temp
        }
        return rob2
    }
};